Faas docs (#1173)
Macrometacorp/docs: docs/photoniq/faas/index.md

---
sidebar_position: 1
title: Function as a Service (FaaS)
---

> 

Reviewer: ## Overview
PhotonIQ FaaS (Function as a Service) is an edge-computing service that allows you to create and run functions without the extra burden of infrastructure management. Thus, IT teams have the time and flexibility to do more tasks, improving operational efficiency and business growth. 
PhotonIQ FaaS uses the highly distributed [Global Data Mesh](https://www.macrometa.com/platform/global-data-mesh) to deploy and execute functions from the closest point of presence(PoP) to customers, thus reducing latency and improving performance. This makes it an ideal solution for developing and managing high-performing, scalable, and low-latency cloud-native applications. 
Currently, FaaS supports Rust and Python, empowering developers with flexibility and eliminating vendor lock-in during development. 
From data processing to application development, the PhotonIQ FaaS revolutionizes application development by removing the complexities of managing infrastructure, thus improving business agility for deploying fast and scalable global serverless functions for your various business needs.

> PhotonIQ FaaS (Function as a Service) is an edge-computing service that allows you to run code in response to events, without managing the complex infrastructure. Deployed functions are executed from the closest point of presence to customers.

Reviewer 2: @joubert-macrometa , @felipe-lopes-macrometa , @dlozina-macrometa - First attempt at an intro page. Please review.

Reviewer 3: I think this "in response to events" is confusing, because we don't need to wait for events to execute a function

Reviewer 4: This is how AWS lamda is described:

_AWS Lambda is a compute service that lets you run code without provisioning or managing servers.

Lambda runs your code on a high-availability compute infrastructure and performs all of the administration of the compute resources, including server and operating system maintenance, capacity provisioning and automatic scaling, and logging. With Lambda, all you need to do is supply your code in one of the language runtimes that Lambda supports.

You organize your code into Lambda functions. The Lambda service runs your function only when needed and scales automatically. You only pay for the compute time that you consume—there is no charge when your code is not running. _

Reviewer 4: just posting to get another example

Reviewer 4: This is what Azure says:
Azure Functions is an event-driven, serverless compute platform that helps you develop more efficiently using the programming language of your choice.

Reviewer 2: Updated

> PhotonIQ FaaS supports Rust and Python, empowering developers with flexibility and eliminating vendor lock-in. This platform is tailored for applications requiring high performance, rapid scalability, and low latency, making it an ideal solution for modern, cloud-native development.

Reviewer 4: Here you say "supports Rust and Python" but in the text above this you say you support "Rust" and "Javascript".

PhotonIQ FaaS revolutionizes application development by abstracting the complexities of infrastructure management, allowing you to deploy fast and scalable serverless functions globally. Whether you are processing data, integrating services, or building an entire application, PhotonIQ FaaS provides the tools and performance you need to succeed.

## How PhotonIQ FaaS Works

> 

Reviewer: PhotonIQ FaaS employs an event-driven model to trigger and deploy lightweight, high-performance functions in the form of [WebAssembly](https://webassembly.org/) modules in response to an HTTP request made to a specific endpoint. 
This event-driven model helps with effective resource consumption as FaaS only utilizes resources when necessary. Furthermore, the stateless nature of FaaS functions ensures their ability to independently handle multiple requests in parallel, thus improving performance.

> PhotonIQ FaaS allows you to deploy lightweight, high-performance functions in the form of [WebAssembly](https://webassembly.org/) modules. When an HTTP request is made to a specific endpoint, the corresponding worker (function) is triggered to process the request. This event-driven model ensures that resources are utilized only when necessary, and the stateless nature of functions ensures they can independently handle multiple requests in parallel. The platform automatically manages the deployment, scaling, and execution of these functions, allowing developers to focus solely on their code.

Reviewer 4: Didn't we come up with a name for "our" functions?  I forgot it, but it might be worth naming it here.

... the corresponding worker (Photoniq function) is triggered...

I can't remember if we called then Photoniq function.

> PhotonIQ FaaS allows you to deploy lightweight, high-performance functions in the form of Wasm modules. When an HTTP request is made to a specific endpoint, the corresponding worker (function) is triggered to process the request. This event-driven model ensures that resources are utilized only when necessary, and the stateless nature of functions ensures they can independently handle multiple requests in parallel. The platform automatically manages the deployment, scaling, and execution of these functions, allowing developers to focus solely on their code.

Reviewer 3: WebAssembly (with capital A)

## Benefits of PhotonIQ FaaS

- **High Performance**: Leverages the near-native execution speed of WebAssembly, enhancing application responsiveness and efficiency.

> - **Global Scalability**: With built-in geo-replication, functions are replicated across multiple regions to reduce latency and improve the reliability of your applications.

Reviewer: This should be reliability.

- **Developer Flexibility**: Supports multiple languages, providing a broad spectrum of tools and libraries to suit your development needs.
- **Cost Efficiency**: Only charges for the actual compute time and resources used, which optimizes spending and reduces waste.
- **Simplified Management**: Automates many operational tasks such as scaling, deployments, and maintenance, freeing up your time and resources.
- **Enhanced Security**: Executes functions in a sandboxed environment, ensuring isolation and protecting against unauthorized access and attacks.

## Features of PhotonIQ FaaS

> - **Multiple Language Support**: Offers compatibility with languages like Rust, JavaScript, Python, Go, and Ruby.

Reviewer 2: Earlier you said it was only Python and JavaScript. Which is correct?

> - **Multiple Language Support**: Offers compatibility with languages like Rust, JavaScript, Python, Go, and Ruby.

Reviewer 3: Rust and Javascript, the other languages will be supported in the future

- **Automatic Scaling**: Dynamically adjusts the number of function instances based on request volume to handle any load efficiently.
- **Geo-Replication**: Automatically distributes function instances across global locations to minimize response times and maximize availability.
- **Integrated Developer Tools**: Provides tools such as a CLI for deployment and management, and a dashboard for monitoring function metrics and logs.
- **Comprehensive Logging and Metrics**: Captures detailed logs and performance metrics to help diagnose issues and optimize function behavior.
- **Advanced Function Capabilities**: Supports HTTP requests, interactions with various Macrometa GDN stores, environment variables, and more.

> - **Secure Execution**: Ensures secure communications and robust security controls for function execution and data handling.

Reviewer: What are some security controls that have been put in place? Can we list some, apart from execution in sandboxes?

---
sidebar_position: 1
title: Function as a Service (FaaS)
---

> Function as a Service (FaaS) is a edge-computing service that allows customers to run code in response to events, without managing the complex infrastructure. Deployed functions will be executed from the closest point of presence to customers.

Reviewer 4: ```suggestion
Function as a Service (FaaS) is an edge-computing service that allows customers to run code in response to events, without managing the complex infrastructure. Deployed functions are executed from the closest point of presence to customers.
```
